src/TooltipPopoverWrapper.js

```diff
@@ -0,0 +1,286 @@
+    PropTypes.func,
+    PropTypes.string,
+    PropTypes.object
+  ]),
+  trigger: PropTypes.string,
+  popperClasses: PropTypes.string,
+  componentType: PropTypes.string,
+  classes: PropTypes.string
+};
+
+const DEFAULT_DELAYS = {
+  show: 0,
+  hide: 250
+};
+
+const defaultProps = {
+  isOpen: false,
+  hideArrow: false,
+  placement: 'top',
+  placementPrefix: 'bs-tooltip',
+  delay: DEFAULT_DELAYS,
+  autohide: true,
+  toggle: function () { },
+  componentType: 'tooltip'
+};
+
+class TooltipPopoverWrapper extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.addTargetEvents = this.addTargetEvents.bind(this);
+    this.handleDocumentClick = this.handleDocumentClick.bind(this);
+    this.removeTargetEvents = this.removeTargetEvents.bind(this);
+    this.toggle = this.toggle.bind(this);
+    this.onMouseOverTooltip = this.onMouseOverTooltip.bind(this);
+    this.onMouseLeaveTooltip = this.onMouseLeaveTooltip.bind(this);
+    this.onMouseOverTooltipContent = this.onMouseOverTooltipContent.bind(this);
+    this.onMouseLeaveTooltipContent = this.onMouseLeaveTooltipContent.bind(this);
+    this.show = this.show.bind(this);
+    this.hide = this.hide.bind(this);
+    this.onEscKeyDown = this.onEscKeyDown.bind(this);
+  }
+
+  componentDidMount() {
+    this._target = getTarget(this.props.target);
+    this.addTargetEvents();
+  }
+
+  componentWillUnmount() {
+    this.removeTargetEvents();
+  }
+
+  onMouseOverTooltip(e) {
+    if (this._hideTimeout) {
+      this.clearHideTimeout();
+    }
+    this._showTimeout = setTimeout(this.show.bind(this, e), this.getDelay('show'));
+  }
+
+  onMouseLeaveTooltip(e) {
+    if (this._showTimeout) {
+      this.clearShowTimeout();
+    }
+    this._hideTimeout = setTimeout(this.hide.bind(this, e), this.getDelay('hide'));
+  }
+
+  onMouseOverTooltipContent() {
+    if (this.props.autohide) {
+      return;
+    }
+    if (this._hideTimeout) {
+      this.clearHideTimeout();
+    }
+  }
+
+  onMouseLeaveTooltipContent(e) {
+    if (this.props.autohide) {
+      return;
+    }
+    if (this._showTimeout) {
+      this.clearShowTimeout();
+    }
+    e.persist();
+    this._hideTimeout = setTimeout(this.hide.bind(this, e), this.getDelay('hide'));
+  }
+
+  onEscKeyDown(e) {
+    if (e.key === 'Escape') {
+      this.hide(e);
+    }
+  }
+
+  getDelay(key) {
+    const { delay } = this.props;
+    if (typeof delay === 'object') {
+      return isNaN(delay[key]) ? DEFAULT_DELAYS[key] : delay[key];
+    }
+    return delay;
+  }
+
+  show(e) {
+    if (!this.props.isOpen) {
+      this.clearShowTimeout();
+      this.toggle(e);
+    }
+  }
+
+  hide(e) {
+    if (this.props.isOpen) {
+      this.clearHideTimeout();
+      this.toggle(e);
+    }
+  }
+
+  clearShowTimeout() {
+    clearTimeout(this._showTimeout);
+    this._showTimeout = undefined;
+  }
+
+  clearHideTimeout() {
+    clearTimeout(this._hideTimeout);
+    this._hideTimeout = undefined;
+  }
+
+  handleDocumentClick(e) {
```

Comment:
I think the logic between tooltip and popover is really the same but it was just implemented differently in the original components. It looks like they do the opposite checks to perform the opposite action.

There is some extra stuff on the second condition more specific to `this._popover`, but it is just checking that the content is not within the popover. A similar check is probably needed for tooltip or made into another prop to control clicks within the "add element".

The goal is for this to be as generic as possible.

Reply:
This is done by passing a `ref` as `prop` down to the `div` inside, right? Since `Tooltip` is providing `innerRef` as a `prop`, how can I do the same for `Tooltip` (since the function can change from the parent created by the user)?